jupyterhub_saml_auth/authenticator.py

```diff
@@ -0,0 +1,108 @@
+from jupyterhub.auth import Authenticator
+from jupyterhub.utils import url_path_join
+from traitlets import Unicode, validate, Set, Callable
+import os
+
+from .handlers import (
+    ACSHandler,
+    MetadataHandler,
+    SamlLoginHandler,
+    SamlLogoutHandler,
+)
+
+
+class SAMLAuthenticator(Authenticator):
+
+    acs_handler = ACSHandler
+    metadata_handler = MetadataHandler
+    login_handler = SamlLoginHandler
+    logout_handler = SamlLogoutHandler
+
+    session_cookie_names = Set(
+        help='''
+        Cookie names for managing SAML session that would be cleared upon
+        logout
+        ''',
+        config=True
+    )
+
+    saml_settings_path = Unicode(
+        default_value='/etc/saml',
+        config=True,
+        help='''
+        A filepath to the location of saml settings required for python3-saml.
+        Contents of this directory should include settings.json,
+        advanced_settings.json, and a cert/ directory if applicable
+        '''
+    )
+
+    extract_username = Callable(
+        help='''
+        Extract the username from the attributes returned by the IdP.
+
+        1. the ACSHandler instance if needed
+        2. Takes in a dict with the attributes, must return a
+        username as a string
+        ''',
+        config=True
+    )
+
+    login_service = Unicode(
+        os.environ.get('LOGIN_SERVICE', 'SSO'),
+        config=True,
+        help='''
+        Hosted domain string, e.g. your IdP
+        ''',
+    )
+
+    @validate('saml_settings_path')
```

Comment:
i like the usage of the validate traitlet, could we review traitlets during our code review?